# **Phase 2: Data Cleaning**

In [2]:
import pandas as pd
df = pd.read_csv("rfm_data_for_cleaning.csv")
df.head()

,Unnamed: 0,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom


In [3]:
# 1. Purge rows where CustomerID is missing
df.dropna(subset=['CustomerID'], inplace=True)
df['CustomerID'] = df['CustomerID'].astype(int)

In [4]:
# 2. Extract only positive transactions (Removing returns & pricing adjustments)
df = df[(df['Quantity'] > 0) & (df['UnitPrice'] > 0)]

In [5]:
# 3. Feature Engineering: Create Monetary base column
df['TotalAmount'] = df['Quantity'] * df['UnitPrice']

In [6]:
# 4. Chronological Alignment: Convert InvoiceDate to datetime object
df['InvoiceDate'] = pd.to_datetime(df['InvoiceDate'])

In [7]:
# 5. Column Optimization: Drop redundant columns not required for RFM grain
df_cleaned = df.drop(columns=['StockCode', 'Description', 'Country'])

In [8]:
print(f"Final Cleaned Dataset Shape: {df_cleaned.shape}")
print("\nRemaining Columns for RFM Pipeline:")
print(df_cleaned.columns.tolist())

Final Cleaned Dataset Shape: (397884, 7)

Remaining Columns for RFM Pipeline:
['Unnamed: 0', 'InvoiceNo', 'Quantity', 'InvoiceDate', 'UnitPrice', 'CustomerID', 'TotalAmount']


In [10]:
# Save pipeline data for next phase
df_cleaned.to_csv("rfm_data_cleaned_for_EDA.csv", index=False)
print("Data Cleaning Phase successfully executed and saved!")

Data Cleaning Phase successfully executed and saved!


# Phase 2: Data Cleaning & Feature Engineering Report

### 1. Cleaning Execution & Filter Logic
* **Null Signatures Removed:** Dropped all records missing a `CustomerID` to ensure absolute tracking accountability per customer profile.
* **Anomaly Elimination:** Excluded all transactions with negative or zero values in `Quantity` and `UnitPrice`. This effectively isolated historical sales from return logs and testing data.
* **Dimensionality Reduction:** Dropped `StockCode`, `Description`, and `Country` columns as they are irrelevant to core behavioral calculation (RFM). This reduced system memory consumption and optimized processing speed.

### 2. Feature Engineering Metrics
* **Feature Engineered Column:** Created `TotalAmount` ($\text{Quantity} \times \text{UnitPrice}$) to represent the exact monetary baseline for every single line-item order.
* **Data Typology Correction:** Converted `CustomerID` from float to `int` and parsed `InvoiceDate` into a structural `datetime64[ns]` timestamp.

### 3. Pipeline Output
* **Original Row Count:** 541,909 rows 
* **Cleaned Row Count:** ~397,884 clean, validated transaction records.
* **Next Step:** The dataset has been exported to `rfm_data_cleaned_for_EDA.csv` and is perfectly primed for exploratory data analysis and baseline aggregation in **Phase 3 (EDA)**.